// nextein/dist/components/post.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Content = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _hoistNonReactStatics = require('hoist-non-react-statics');

var _hoistNonReactStatics2 = _interopRequireDefault(_hoistNonReactStatics);

var _unified = require('unified');

var _unified2 = _interopRequireDefault(_unified);

var _remarkParse = require('remark-parse');

var _remarkParse2 = _interopRequireDefault(_remarkParse);

var _remarkRehype = require('remark-rehype');

var _remarkRehype2 = _interopRequireDefault(_remarkRehype);

var _rehypeReact = require('rehype-react');

var _rehypeReact2 = _interopRequireDefault(_rehypeReact);

var _rehypeRaw = require('rehype-raw');

var _rehypeRaw2 = _interopRequireDefault(_rehypeRaw);

var _unistUtilSelect = require('unist-util-select');

var _unistUtilSelect2 = _interopRequireDefault(_unistUtilSelect);

var _load = require('../entries/load');

var _utils = require('./utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var extractExcerpt = function extractExcerpt(excerpt) {
  if (!excerpt) {
    return;
  }

  var selector = typeof excerpt === 'string' ? excerpt : ':root > paragraph:first-child';

  return function () {
    return (/* attacher */function (tree) {
        /* transformer */
        tree.children = (0, _unistUtilSelect2.default)(tree, selector);
        return tree;
      }
    );
  };
};

var toReact = function toReact(_ref) {
  var content = _ref.content,
      excerpt = _ref.excerpt,
      renderers = _ref.renderers,
      _ref$prefix = _ref.prefix,
      prefix = _ref$prefix === undefined ? 'entry-' : _ref$prefix;
  return (0, _unified2.default)().use(_remarkParse2.default).use(extractExcerpt(excerpt)).use(_remarkRehype2.default, { allowDangerousHTML: true }).use(_rehypeRaw2.default).use(_rehypeReact2.default, {
    createElement: _react2.default.createElement,
    prefix: prefix,
    components: renderers
  }).processSync(content).contents;
};

var Content = function Content(props) {
  var content = props.content,
      excerpt = props.excerpt,
      renderers = props.renderers,
      data = props.data,
      prefix = props.prefix,
      componentProps = _objectWithoutProperties(props, ['content', 'excerpt', 'renderers', 'data', 'prefix']);

  var cmp = toReact({ content: content, excerpt: excerpt, renderers: renderers, prefix: prefix });
  var cmpProps = cmp.props;


  return _extends({}, cmp, {
    props: _extends({}, cmpProps, componentProps)
  });
};

exports.Content = Content;

exports.default = function (WrappedComponent) {
  var _class, _temp;

  var displayName = (0, _utils.getDisplayName)(WrappedComponent);

  return (0, _hoistNonReactStatics2.default)((_temp = _class = function (_Component) {
    _inherits(_class, _Component);

    function _class() {
      _classCallCheck(this, _class);

      return _possibleConstructorReturn(this, (_class.__proto__ || Object.getPrototypeOf(_class)).apply(this, arguments));
    }

    _createClass(_class, [{
      key: 'render',
      value: function render() {
        var props = this.props;


        return _react2.default.createElement(WrappedComponent, props);
      }
    }], [{
      key: 'getInitialProps',
      value: async function getInitialProps() {
        var wrappedInitial = WrappedComponent.getInitialProps;

        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
          args[_key] = arguments[_key];
        }

        var wrapped = wrappedInitial ? await wrappedInitial.apply(undefined, args) : {};
        var _args$0$query = args[0].query,
            query = _args$0$query === undefined ? {} : _args$0$query;
        var _entry = query._entry;

        var post = _entry ? await (0, _load.byFileName)(_entry) : undefined;

        return _extends({}, wrapped, {
          post: post
        });
      }
    }]);

    return _class;
  }(_react.Component), _class.displayName = 'WithPost(' + displayName + ')', _temp), WrappedComponent, { 'getInitialProps': true });
};
